core/builtin.py

```diff
@@ -485,18 +485,37 @@ def Cd(argv, mem):
 
 def Pushd(argv, dir_stack):
   dir_stack.append(os.getcwd())
+  num_args = len(argv)
+
+  if num_args <= 0:
+    log("pushd: no other directory")
+    return 1
+  elif num_args > 1:
+    log("pushd: too many arguments")
+
   dest_dir = argv[0]
-  os.chdir(dest_dir)  # TODO: error checking
+  if os.path.isdir(dest_dir):
+    os.chdir(dest_dir)
+  else:
+    log("pushd: " + dest_dir + ": No such file or directory")
+    return 1
+
   return 0
 
 
 def Popd(argv, dir_stack):
   try:
     dest_dir = dir_stack.pop()
   except IndexError:
-    log('popd: directory stack is empty')
+    log("popd: directory stack is empty")
+    return 1
+
+  if os.path.isdir(dest_dir):
```

Comment:
Please change this test and the one above to try the chdir and catch OSError.

I don't think `isdir()` matches the conditions where `chdir()` fails (symlink to dir, etc.?).  So it's cleaner to only show an error when the actual chdir() call fails.

Reply:
Would you prefer that the error message printed when catching OSError is the same as the error message produced in bash? Or would you prefer that I follow the same error message that you print in the Cd() builtin? (which is: `util.error("cd %r: %s", dest_dir, os.strerror(e.errno))`)

Reply:
Yes good question.  The cd style is better, because it prints the actual error from libc.

Actually, bash probably does the same thing -- it doesn't always print "No such file or directory".  It will also print "Permission denied", for example, if that is the `errno` code.

Reply:
Ok, sounds good. I'll stick with the Cd() style